AP_InertialSensor: Murata SCH16T IMU (Integrated FIFO)

ZeroOne-Aero opened the pull request:
This is an improved Murata SCH16T-K10 IMU, which integrates FIFO functionality through an FPGA and reads sensor data via callback

tridge commented:
@ZeroOne-Aero there are a few build errors (see CI logs with clang for example)
Can you supply a flight log, with this IMU and another IMU, with LOG_DISARMED=1 and LOG_REPLAY=1 ?

ZeroOne-Aero commented:
@peterbarker Hello Peter, if you have time, could you please take a look at the flight logs we submitted?

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_SCH16T_Fifo.cpp:
```diff
@@ -0,0 +1,760 @@
+/*Add commentMore actions
```
```suggestion
/*
```

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_ZeroOnex.cpp:
```diff
@@ -0,0 +1,760 @@
+uint8_t direct_mode = CtrlMode_Direct;
+uint8_t fifo_enable = 0;
+uint8_t fifo_cmd_num = 17;
+uint8_t fifo_baudrate = 32;
```
member variables, please

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_SCH16T_Fifo.cpp:
```diff
@@ -0,0 +1,760 @@
+
+void AP_InertialSensor_ZeroOnex::seneor_ctrl(uint8_t fifo_rst) 
+{
+    uint32_t reg;
```
```suggestion
```

declare variables where their value is first set

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_SCH16T_Fifo.cpp:
```diff
@@ -0,0 +1,760 @@
+    uint8_t wbuf[8];
+    uint8_t ptr = 0;
+
+    reg = (direct_mode & 1) << CTRL_Shift_Mode;
```
```suggestion
    uint32_t reg = (direct_mode & 1) << CTRL_Shift_Mode;
```

tridge commented on libraries/AP_InertialSensor/AP_InertialSensor_ZeroOnex.cpp:
```diff
@@ -0,0 +1,760 @@
+        GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "failed,wreg is:%d,rreg is:%d",wreg,rreg);
+        return 0;
+    }
+    GCS_SEND_TEXT(MAV_SEVERITY_INFO, "success,wreg is:%d,rreg is:%d",wreg,rreg);
```
too many verbose status texts, please reduce for production use, or put behind an #if  (this is not the only one, check all GCS_SEND_TEXT)

tridge commented on libraries/AP_InertialSensor/AP_InertialSensor_ZeroOnex.cpp:
```diff
@@ -0,0 +1,760 @@
+{
+    _sch16t_read_status(STAT_SUM);
+    _sensor_status.summary      = _sch16t_read_status(STAT_SUM);
+    _sensor_status.summary      = _sch16t_read_status(STAT_SUM_SAT);
```
same variable twice?

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_ZeroOnex.cpp:
```diff
@@ -0,0 +1,760 @@
+    wbuf[ptr++] = reg & 0xff;
+    wbuf[ptr++] = (reg >> 8) & 0xff;
+    wbuf[ptr++] = (reg >> 16) & 0xff;
+    wbuf[ptr++] = (reg >> 24) & 0xff;
```
Consider using a union of the buffer and a structure

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_ZeroOnex.cpp:
```diff
@@ -0,0 +1,760 @@
+    seneor_ctrl(1);
+}
+
+static uint8_t gen_crc8(uint8_t* data) {
```
Do we have a common crc function for this in `AP_Math/crc.h`

ZeroOne-Aero replied:
no,there are some differences between the crc function we have commited and common crc function  in AP_Math/crc.h

tridge commented:
> This is the flight log of our X6 as the flight controller.

thanks, the log looks reasonable to me

ZeroOne-Aero commented:
@tridge Thank you for your suggestion. I will modify the code carefully before submitting it. This is the flight log with EK3_IMU_MASK, LOG_DISARMED=1, and LOG_REPLAY=1 set. SCH16T is the third IMU.
https://github.com/ZeroOne-Aero/Flight-Logs

vidma commented:
why not to use DRDY interrupt?

peterbarker reviewed (changes requested):
Sorry, partial review only

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_SCH16T_Fifo.h:
```diff
@@ -0,0 +1,176 @@
+    uint8_t accel_instance {};
+    uint8_t gyro_instance = {};
+    float accel_scale {};
+    float gyro_scale {};
```
Do not zero any memory in here, it wastes flash and isn't a pattern we want in ArduPilot

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor.cpp:
```diff
@@ -42,6 +42,7 @@
 #include "AP_InertialSensor_ASM330.h"
 #include "AP_InertialSensor_ADIS16607.h"
 #include <AP_Scheduler/AP_Scheduler.h>
+#include "AP_InertialSensor_SCH16T_Fifo.h"
```
I think you renamed this driver?

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_Backend.h:
```diff
@@ -155,6 +155,7 @@ class AP_InertialSensor_Backend
         DEVTYPE_INS_LSM6DSV  = 0x3E,
         DEVTYPE_INS_ASM330   = 0x3F,
         DEVTYPE_INS_ADIS16607 = 0x40,
+        DEVTYPE_INS_SCH16T = 0x41,
```
Need to be more specific in here - this is more than just the SCH16T

Also note that this ID may be consumed by some other driver before this is merged - we'll see what gets in first.

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_SCH16T_Fifo.cpp:
```diff
@@ -0,0 +1,741 @@
+
+#include <AP_Math/AP_Math.h>
+
+#include "AP_InertialSensor_SCH16T_Fifo.h"
```
```suggestion
#include "AP_InertialSensor_SCH16T_FIFO.h"
```

... but I do suggest you come up with a name for your new "sensor" with the FPGA in front of the actual SCH16T.  I think we were actually OK with "ZeroOne" being in the AP_InertialSensor_* name,

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_SCH16T_Fifo.cpp:
```diff
@@ -0,0 +1,741 @@
+
+    case State::Validate: {
+            read_status_registers(); // Read all status registers twice
+            read_status_registers();
```
I can see you're reading everything twice here.

Explain why, not what

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_SCH16T_Fifo.cpp:
```diff
@@ -0,0 +1,741 @@
+    } // end switch/case
+}
+
+bool AP_InertialSensor_SCH16T_Fifo::collect_and_publish() {
```
```suggestion
bool AP_InertialSensor_SCH16T_Fifo::collect_and_publish()
{
```

similarly elsewhere - required style

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_SCH16T_Fifo.cpp:
```diff
@@ -0,0 +1,741 @@
+    _rotate_and_correct_gyro(gyro_instance, gyro);
+    _notify_new_gyro_raw_sample(gyro_instance, gyro);
+
+    _publish_temperature(accel_instance, float(data.temp)/100.f);
```
Bad indent

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_SCH16T_Fifo.cpp:
```diff
@@ -0,0 +1,741 @@
+    fpga_read_fifo8(Addr_W8_SensorValueFifo8, buff, 24);//one packet contains 24 bytes
+    fpga_write_reg8(Addr_RW8_SensorFifoCtrl, 0x01);//notice the fpga to load next packet
+
+    if (buff[0] & 0x80) data.gyro_x = SPI32BITCONVERT2_20BIT(((buff[1] & 0xf) << 16) | (buff[2] << 8) | buff[3]);
```
```suggestion
    if (buff[0] & 0x80) { data.gyro_x = SPI32BITCONVERT2_20BIT(((buff[1] & 0xf) << 16) | (buff[2] << 8) | buff[3]) };
```

Similarly everywhere else.  Braces are strictly required.  See our [style guide](https://ardupilot.org/dev/docs/style-guide.html) for more information

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_SCH16T_Fifo.cpp:
```diff
@@ -0,0 +1,741 @@
+}
+
+void AP_InertialSensor_SCH16T_Fifo::configure_registers() {
+    uint32_t reg_value;
```
Declare variables as late as possible.

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_SCH16T_Fifo.cpp:
```diff
@@ -0,0 +1,741 @@
+
+bool AP_InertialSensor_SCH16T_Fifo::validate_sensor_status() {
+    auto &s = _sensor_status;
+    uint16_t values[] = { s.summary, s.saturation, s.common, s.rate_common, s.rate_x, s.rate_y, s.rate_z, s.acc_x, s.acc_y, s.acc_z };
```
```suggestion
    const uint16_t values[] = { s.summary, s.saturation, s.common, s.rate_common, s.rate_x, s.rate_y, s.rate_z, s.acc_x, s.acc_y, s.acc_z };
```

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_SCH16T_Fifo.cpp:
```diff
@@ -0,0 +1,741 @@
+
+bool AP_InertialSensor_SCH16T_Fifo::validate_register_configuration() {
+    uint32_t value;
+    bool result = true;
```
```suggestion
```

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_ZeroOne_FPGA_SCH16T.cpp:
```diff
@@ -0,0 +1,741 @@
+        register_read(r.addr,0); // double read, wasteful but makes the code cleaner, not high rate so doesn't matter anyway
+        register_read(r.addr,&value); 
+
+        if (value != r.value) {
```
```suggestion
        uint32_t value;
        if (value != r.value) {
```

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_SCH16T_Fifo.cpp:
```diff
@@ -0,0 +1,741 @@
+        register_read(r.addr,&value); 
+
+        if (value != r.value) {
+            result = false;
```
```suggestion
            return false;
```

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_SCH16T_Fifo.cpp:
```diff
@@ -0,0 +1,741 @@
+        }
+    }
+
+    return result;
```
```suggestion
    return true;
```

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_SCH16T_Fifo.cpp:
```diff
@@ -0,0 +1,741 @@
+
+    for (auto &r : _registers) {
+        register_read(r.addr,0); // double read, wasteful but makes the code cleaner, not high rate so doesn't matter anyway
+        register_read(r.addr,&value); 
```
Why?

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_SCH16T_Fifo.cpp:
```diff
@@ -0,0 +1,741 @@
+uint8_t AP_InertialSensor_SCH16T_Fifo::wait_direct_busy(void) {
+    uint16_t timeout = 0x200;
+    while (timeout) {
+        if (!(fpga_read_reg16(Addr_RW16_DirectSpiCtrl) & 1)) return 1;
```
braces are always required

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_SCH16T_Fifo.cpp:
```diff
@@ -0,0 +1,741 @@
+    dev->transfer_fullduplex(frame.raw,5+size);
+    for(uint8_t i=0;i<size;i++)
+    {
+        *(value+i) =frame.fifo.data[i+1];
```
```suggestion
        *(value+i) = frame.fifo.data[i+1];
```

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_SCH16T_Fifo.cpp:
```diff
@@ -0,0 +1,741 @@
+    for(uint8_t i=1;i<size;i++)
+    {
+        frame.fifo.data[i] = *(value+i);
+    }
```
This looks like a memcpy?

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_SCH16T_Fifo.cpp:
```diff
@@ -0,0 +1,741 @@
+    frame.reg.cmd= FuncBit_Write | FuncBit_Reg | FuncBit_Bit8;
+    frame.reg.reg_addr_l= reg & 0xff;
+    frame.reg.reg_addr_h= (reg >> 8) & 0xff;
+    frame.reg.data[0]= value;
```
```suggestion
    frame.reg.reg_addr_l= reg & 0xff;
    frame.reg.reg_addr_h = (reg >> 8) & 0xff;
    frame.reg.data[0] = value;
```

peterbarker replied:
also consider initialising `frame` as part of its declaration

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_SCH16T_Fifo.cpp:
```diff
@@ -0,0 +1,741 @@
+    frame.reg.reg_addr_l= reg & 0xff;
+    frame.reg.reg_addr_h= (reg >> 8) & 0xff;
+    frame.reg.data[0]= SEPARATOR_BYTE;
+    frame.reg.data[1]= DUMMY_BYTE;
```
Consider initialising `frame` as part of its declaration.  

Similarly elsewhere.  It helps catches bugs!


```suggestion
    frame.reg.cmd = FuncBit_Read | FuncBit_Reg | FuncBit_Bit8;
    frame.reg.reg_addr_l= reg & 0xff;
    frame.reg.reg_addr_h = (reg >> 8) & 0xff;
    frame.reg.data[0] = SEPARATOR_BYTE;
    frame.reg.data[1] = DUMMY_BYTE;
```


similar whitespace iussues throughout file, fix them all

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_SCH16T_Fifo.cpp:
```diff
@@ -0,0 +1,741 @@
+void AP_InertialSensor_SCH16T_Fifo::fpga_write_reg16(uint16_t reg, uint16_t value) {
+    FpgaFrame frame;
+
+    frame.reg.cmd= FuncBit_Write | FuncBit_Reg | FuncBit_Bit16;
```
It looks like you should be able to factor out a method which takes the command - there's a bunch of things in here that look very similar.

Maybe `fill_fpga_frame(uint8_t cmd, uint16_t reg, uint16_value, uint16_t *data, uint8_t datalen)`

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_SCH16T_Fifo.cpp:
```diff
@@ -0,0 +1,741 @@
+    fpga_write_reg16(Addr_RW16_TestReg, wreg);
+    rreg = fpga_read_reg16(Addr_RW16_TestReg);
+    if (wreg != rreg) 
+    {
```
```suggestion
    if (wreg != rreg) {
```

again, see style guide

Commit:
AP_InertialSensor: add FPGA_SCH16T_K10 FIFO driver

ZeroOne-Aero commented:
@peterbarker We have modified the code according to your suggestions, and added comments next to the code to address the code duplication issues. Please review it again to see if there are any further problems.

peterbarker commented on libraries/AP_InertialSensor/AP_InertialSensor_Backend.h:
```diff
@@ -155,6 +155,7 @@ class AP_InertialSensor_Backend
         DEVTYPE_INS_LSM6DSV  = 0x3E,
         DEVTYPE_INS_ASM330   = 0x3F,
         DEVTYPE_INS_ADIS16607 = 0x40,
+        DEVTYPE_INS_SCH16T_K10 = 0x50,
```
```suggestion
        DEVTYPE_INS_ZEROONE_FPGA_SCH16T = 0x50,
```

peterbarker replied:
I've made this fix.

tridge commented:
can reference https://github.com/ArduPilot/Schematics/tree/master/ZeroOne/FPGA/SCH16T/ef2 directly in the comments in the driver

tridge commented on libraries/AP_InertialSensor/AP_InertialSensor_ZeroOne_FPGA_SCH16T.cpp:
```diff
@@ -0,0 +1,788 @@
+bool AP_InertialSensor_ZeroOne_FPGA_SCH16T::read_product_id()
+{
+    uint32_t comp_id = 0, asic_id = 0;
+    register_read(COMP_ID, 0);//the FPGA doesn't process the first instruction,it only starts processing after receiving the second instruction,resulting in the same command being sent twice.
```
should be nullptr

tridge reviewed (changes requested):
can we get a log? the existing log link is stale

peterbarker reviewed (dismissed):
We'd like to see a flight log from this final code before we merge it; claude is claiming an 16x scale error on the gyros., so we'd like to see it working against some other known gyro in a log.

peterbarker commented:
If the log looks good and a final pass from Peter's fine then we can merge this

ZeroOne-Aero commented:
https://github.com/ZeroOne-Aero/Flight_Logs/tree/main
The two files named ZeroOne_SCH16T_FIFO_251218 need to be decompressed together.

tridge commented:
@ZeroOne-Aero which of the IMUs is the new one? This is what devid gives me on that log:
```
BARO1_DEVID: bus_type:I2C(1)  bus:0 address:100(0x64) devtype:16(0x10) BARO_ICP201XX (1074177)
BARO2_DEVID: bus_type:I2C(1)  bus:2 address:99(0x63) devtype:16(0x10) BARO_ICP201XX (1073937)
COMPASS_DEV_ID: bus_type:UAVCAN(3)  bus:0 address:123(0x7b) devtype:1(0x1) UAVCAN (97027)
COMPASS_DEV_ID2: bus_type:I2C(1)  bus:0 address:32(0x20) devtype:17(0x11) RM3100 (1122305)
COMPASS_PRIO1_ID: bus_type:UAVCAN(3)  bus:0 address:123(0x7b) devtype:1(0x1) UAVCAN (97027)
COMPASS_PRIO2_ID: bus_type:I2C(1)  bus:0 address:32(0x20) devtype:17(0x11) RM3100 (1122305)
INS_ACC2_ID: bus_type:SPI(2)  bus:1 address:1(0x1) devtype:61(0x3d) INS_IIM42653 (3997962)
INS_ACC3_ID: bus_type:SPI(2)  bus:2 address:1(0x1) devtype:62(0x3e) INS_LSM6DSV (4063506)
INS_ACC_ID: bus_type:SPI(2)  bus:3 address:1(0x1) devtype:61(0x3d) INS_IIM42653 (3997978)
INS_GYR2_ID: bus_type:SPI(2)  bus:1 address:1(0x1) devtype:61(0x3d) INS_IIM42653 (3997962)
INS_GYR3_ID: bus_type:SPI(2)  bus:2 address:1(0x1) devtype:62(0x3e) INS_LSM6DSV (4063506)
INS_GYR_ID: bus_type:SPI(2)  bus:3 address:1(0x1) devtype:61(0x3d) INS_IIM42653 (3997978)
```

tridge commented:
@ZeroOne-Aero looking at the hwdef.dat in the log it seems the 3rd IMU is the new one. The IMU[2].GHz and IMU[2].AHz is only 1468Hz - is this expected? Seems quite low
ahh, yes, I see it is targetting 1475Hz, so 1468 is close. Can it do a higher rate? Useful for some vehicle types
